### ISSUE TITLE
[RFC] Add `modeling_xxx_fusion.py` to support kernel fusion

### ISSUE THREAD

**Issue by hyunwoongko** [huggingface/transformers#13845]
## Introduction
I am an engineer currently working on 3D model parallelism for transformers. When the tensor model parallelism (https://github.com/huggingface/transformers/pull/13726) is done, I am going to introduce [kernel fusion](https://ieeexplore.ieee.org/document/5724850) feature to transformers. For this I want to create a new modeling file called `modeling_xxx_fusion.py`. This work is currently being discussed with  me and @RezaYazdaniAminabadi (DeepSpeed team). 

## Kernel fusion API
```python
from transformers import BertForMaskedLM

# create model
model = BertForMaskedLM.from_pretrained("bert-base-cased")

# fuse all modules (all arguments is True as default)
# `fuse_modules` is function level fusion, It supports a wide variety of models.
model.fuse_modules()  

# fuse selective modules
model.fuse_modules(
    word_embedding=True,
    scale_mask_softmax=True,
    layer_norm=True,
    bias_gelu=True,
    bias_dropout_residual=False,
)

# fuse whole layers 
# default is `None` -> `not self.training` of `torch.nn.Module`
# `fuse_layers` is block level (attention mlp) fusion, only a few models are supported.
model.fuse_layers(inference=None)

# fuse whole layers for inference
model.fuse_layers(inference=True)

# fuse whole layers for training
model.fuse_layers(inference=False)
```

## Implementation
The internal module of each model will be re-implemented using kernel fusion method, and the existed module will be replaced with the fused module. The following example is an example of `BertOutput(nn.Module)`.

```python
# transformers/models/bert/modeling_bert.py

class BertOutput(nn.Module):
        def __init__(self, config):
              super().__init__()
              self.dense = nn.Linear(config.intermediate_size, config.hidden_size)
              self.LayerNorm = nn.LayerNorm(config.hidden_size, eps=config.layer_norm_eps)
              self.dropout = nn.Dropout(config.hidden_dropout_prob)

        def forward(self, hidden_states, input_tensor):
                hidden_states = self.dense(hidden_states)
                hidden_states = self.dropout(hidden_states)
                hidden_states = self.LayerNorm(hidden_states + input_tensor)
                return hidden_states
```
```python
# transformers/models/bert/modeling_bert_fusion.py

class FusedBertOutput(BertOutput):
        def forward(self, hidden_states, input_tensor):
                hidden_states = hidden_states @ self.dense.weight.t()
                hidden_states = FusedBiasDropoutResidual.apply(hidden_states, self.dense.bias, input_tensor)
                hidden_states = FusedLayerNorm.apply(hidden_states, self.LayerNorm.weight, self.LayerNorm.bias)
                return hidden_states
```
When the user calls the `fuse_modules()` method, the kernel fusion engine finds `BertOutput` and replaces it with `FusedBertOutput`. and user callse `fused_layers` method, engine finds `BertLayer` and replcases id with `FusedBertLayer`. This is the method that `parallelformers` parallelized transformers models flexibly, and the `deepspeed inference` also supports kernel fusion in this way. 

However, the current version of `deepspeed inference` fuses the entire transformer layer, so the supported models are very limited. For example, bigbird requires random attention mechanism. in this case random attention must be implemented in the custom cuda kernel. However, because the number of models is so large, it is impossible to implement them all. So I propose a flexible way to fuse the kernel on a per-function. This is a strategy of choice and concentration. The area that can be fused performs fusion, and the area that can not be fused uses the torch's default module.

```python
# kernel_fusion_utils.py

class KernelFusionMixin(object):
    
    def fuse_modules(...):
        assert self._is_fusionable, "error message"
        ... implementation ...

    def fuse_layers(...)
        assert self._is_fusionable, "error message"
        ... implementation ...

# modeling_utils.py

class PreTrainedModel(..., KernelFusionMixin):
    _is_parallelizable = ...
    _is_able_to_fuse = False. # <--- Only models that implement fusion have True.
```

This is a draft. The API can be changed at any time. I look forward to feedback.

cc. @Stas00 @RezaYazdaniAminabadi @Sylvain

**Comment by stas00** [comment]
Looks like an awesome plan, @hyunwoongko! So far looks good.

I'd just suggest `s/_is_able_to_fuse/_is_fusable/` to use the same style as `_is_parallelizable` if that's what we use. But this is a minor details and we can rename easily later.

**Comment by hyunwoongko** [comment]
> I'd just suggest s/_is_able_to_fuse/_is_fusable/ to use the same style as _is_parallelizable if that's what we use. But this is a minor details and we can rename easily later.

@stas00 You're right. It's because I'm not good at English (I didn't know there was a word `fusable`. lol)

**Comment by stas00** [comment]
In software we often create new words anyway, so as long as the composite of 2 words makes sense it works for our purposes.

Latin languages are all based on the combination of a root with prefx/postfix, so if it's not there already - create one ;)

**Comment by hyunwoongko** [comment]
@stas00 @siddk I'm going to take a look at functorch this weekend and see how we can combine them to improve performance.

**Comment by stas00** [comment]
the tricky part is that it's tied to torch's version, e.g. I had to use pytorch-nightly to get it to work. Otherwise you can only use pt-1.10.0 which is not the latest release (1.10.1 is). To keep up with details please subscribe to: https://github.com/huggingface/transformers/pull/15264

**Comment by Chillee** [comment]
@stas00 We'll be cutting a branch that works with PyTorch 1.11.0, and to be honest, I don't think it'd be that hard to cut a release for 1.10.1 now either.

So, I think the issues with user setup are not that difficult to resolve.

**Comment by tatami-galaxy** [comment]
This looks really useful. Is there a more recent update on this somewhere?
